fix unroll in successive unflatten

avikchaudhuri opened the pull request:
We use `nn_module_stack` in `unflatten` to recognize when module calls begin and end. However the current format is not sufficient to detect module call boundaries when we have successive calls to the same module, because the successive instructions (end of one call, begin of next call) have the same `nn_module_stack`. This causes us to effectively "unroll" successive calls to a single call. This can cause problems when preserving module call signatures because the outputs of the successive calls might be concatenated in the single call.

Previously we introduced the concept of a "call index" to generate multiple graphs when unflattening, one per call. This PR pushes this concept into `nn_module_stack` itself. Thus entries in `nn_module_stack` go from `(fqn, type)` to `(fqn, type, call_index)`.

Note that we still do not have the ability to preserve module call signatures for multiple calls to the same module. But now instead of randomly crashing we give a proper error. OTOH when not preserving module call signatures we simply generate multiple calls, each with its own graph, possibly deduplicated, matching what we would do for non-successive calls.

Differential Revision: D64014936


cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

angelayi commented on torch/_dynamo/variables/nn_module.py:
```diff
@@ -82,8 +82,14 @@ def convert_to_fake(x):
+        tx.nn_module_stack[module_key] = (
+            fully_qualified_name,
+            mod.__class__,
+            num_calls,
```
I wonder if instead of adding another field to the nn_module_stack (which should require a schema minor version bump), we could just modify the key to also contain the num_call? like n_1, n_2..

avikchaudhuri replied:
You mean `module_key` or `fully_qualified_name`?

As you can see, there are lots of places that use `nn_module_stack`, nervous about changing more places. It could be that `module_key` is safest...do you roughly know how important it is?

avikchaudhuri replied:
For schema minor version bump, didn't see any tests fire.

angelayi replied:
I was referring to `module_key`. I'm not too sure how accurate `module_key` needs to be maintained, but skimming the dynamo code, it looks like it's just generated from this [one function](https://www.internalfb.com/code/fbsource/[2d94c2bdacd7]/fbcode/caffe2/torch/_dynamo/output_graph.py?lines=747-760), so it feels like we could just append the `num_call`?

avikchaudhuri replied:
Hmm, I'm not very keen on this idea. Unflatten uses the key as a module id to bucket shared modules; but let's say I can have a local decoder for that format. But the onnx/_internal/fx passes still need to be updated, I think, because they extract the key as "module name" in their metadata structure.

OTOH the utils in ao/quantization seem to only look at values. So does internal AIMP stuff. I won't know what else may rely on keys without trying.

So maybe I can put up a parallel PR that puts this stuff into the key just to check, WDYT?

angelayi replied:
> Unflatten uses the key as a module id to bucket shared modules

Wouldn't fully_qualified_name be a better key to bucket? 

> So maybe I can put up a parallel PR that puts this stuff into the key just to check, WDYT?

Sounds good! Thanks! I'm not opposed to adding another field to the value -- just wanted to see if there was another way to do it.

avikchaudhuri replied:
> Wouldn't fully_qualified_name be a better key to bucket?

`self.n` and `self.p` can point to the same module instance, in which case I think they get the same key (at least in Dynamo).

avikchaudhuri replied:
CI looks green, please review the other PR. Overall that PR makes fewer changes, so I can close this PR if you wish.

avikchaudhuri replied:
Recording offline chat with @angelayi here for posterity. We're going with the alt fix because keys seem less load-bearing than values in `nn_module_stack`, so less chance of breaking unknown workflows.

yiming0416 commented on torch/_export/serde/serialize.py:
```diff
@@ -2199,8 +2200,8 @@ def deserialize_meta_func(serialized_target: str):
-            def import_nn_module_stack(key, path, ty):
-                return key, (path, ty)
+            def import_nn_module_stack(key, path, ty, num_calls):
+                return key, (path, ty, int(num_calls))
```
If someone has a previously serialized EP and calls this new deserializer, looks like they will run into BC issue because `num_calls` doesn't exist in the previously serialized EP?

avikchaudhuri replied:
Yes. Is there precedent in the deserializer for doing something like, fill in default if not present?

yiming0416 replied:
I am not sure about any precedent... But we do need something like this to handle the BC for now. 

Ultimately, we'd like to schematize the `nn_module_stack`, even the whole `metadata`. But maybe that will be done in a separate PR later.

avikchaudhuri replied:
OK, will do.

avikchaudhuri committed:
fix unroll in successive unflatten (#137457)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/137457

Test Plan: added test

Differential Revision: D64014936
